### ISSUE TITLE
Look into moving realm terrain generation into an async environment (https://minetest.gitlab.io/minetest/minetest-namespace-reference/#async-environment).

### ISSUE THREAD

**Issue by lukasgolson** [ubc-minetest-classroom/Minetest_Classroom#166]
A significant part of generating realm terrain involves calculating which nodes should be placed where. The lag caused by this process might be alleviated by passing the LVM data into an async environment for calculations. Retrieving these results, then saving the results in the synchronous environment.

https://minetest.gitlab.io/minetest/minetest-namespace-reference/#async-environment

**Comment by pauldpickell** [comment]
I see `VoxelManip` and `VoxelArea` are available in an async environment. Could we use this approach to place large schematics like UBC campus?

**Comment by lukasgolson** [comment]
> I see `VoxelManip` and `VoxelArea` are available in an async environment. Could we use this approach to place large schematics like UBC campus?

Absolutely! This approach is similar to the one used for the `exschem` and `exschemwe` realm schematic formats. We'd just need to serialize the UBC map using the exschem format to take advantage of it (e.g., load up the UBC map into a realm, and then re-save it using `/realm schematic save <ID> <UBCMapName> exschem`. I'll do some testing to see how this would impact performance.

**Comment by pauldpickell** [comment]
Made some progress towards this using the [mapgen voxel manipulator method](https://minetest.gitlab.io/minetest/lua-voxel-manipulator/#mapgen-voxelmanip-objects), which seems to suit our purpose. We can write and emerge nodes by mapchunk, which is far more efficient than placing/loading schematics. 

Basically, I have added a registration method for bitmap images (similar to how we currently handle schematics) and exposed the image reading functionality of `realterrain`. So the method looks something like `/realm realterrain load UBCV_Campus`, which writes the elements of the UBC campus (and other digital twins, of course) from scratch.

This functionality is now implemented as of [d99988c](https://github.com/ubc-minetest-classroom/minetest_classroom/pull/179/commits/d99988ccb1bfb508bb9e33f647d1fc88596da8f6), however, as of now the `realterrain` biomes are not generated. I think we should utilize our own biomes that are registered with `biomegen` instead (TODO).
